src/libsystemd/sd-device/sd-device.c

```diff
@@ -1429,6 +1429,50 @@ _public_ int sd_device_get_seqnum(sd_device *device, uint64_t *ret) {
         return 0;
 }
 
+int device_get_diskseq(sd_device *device, int fd, bool verify, uint64_t *ret) {
+        const char *s;
+        bool ignore;
+        int r;
+
+        assert(device);
+        assert(fd >= 0);
+
+        /* Accessing the property will likely initialize the device if it's not done. Otherwise we'll ignore
+         * diskseq. */
+        r = sd_device_get_property_value(device, "ID_IGNORE_DISKSEQ", &s);
+        if (r < 0) {
+                if (r != -ENOENT)
+                        return r;
+
+                /* ENOENT means one of the two following cases: a) the device is still not initialized or b)
+                 * the property is not defined. Unlike the latter, the former will ignore diskseq. */
+                ignore = !device->is_initialized;
+
+        } else {
+                r = parse_boolean(s);
+                if (r < 0)
+                        return r;
+
+                ignore = r;
+        }
+
+        uint64_t diskseq = 0;
+        if (!ignore) {
+                r = fd_get_diskseq(fd, &diskseq);
+                if (r < 0)
+                        return r;
+
+                /* Verify that the value returned by the device still matches the one stored in the
+                 * db. */
+                if (verify && diskseq != device->diskseq)
+                        return -ENXIO;
+        }
+        if (ret)
+                *ret = diskseq;
+
+        return 0;
+}
```

Comment:
this function is still weird to me. smells to me this should be two functions. one called device_shall_ignore_diskseq() and the other device_verify_diskseq().

Reply:
It looks better indeed, thanks for the suggestion.